designs/code-generate-image-plugin.md

```diff
@@ -0,0 +1,296 @@
+}
+
+// imageFor{{ resource }} returns the image for the resources
+// belonging to the given {{ resource }} CR name.
+func imageFor{{ resource }}(name string) string {
+	// TODO: this method will return the value of the envvar create to store the image:tag informed 
+}
+
+func (r *{{ resource }}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&cachev1alpha1.{{ resource }}{}).
+		Owns(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+``` 
+
+**Example of the spec for the <kind>_types.go template**
+
+```go
+// {{ resource }}Spec defines the desired state of {{ resource }}
+type {{ resource }}Spec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+    // +kubebuilder:validation:Minimum=1
+	// Size defines the number of {{ resource }} instances
+	Size int32 `json:"size,omitempty"`
+
+    // ImagePullPolicy defines the policy to pull the container images
+	ImagePullPolicy string `json:"image-pull-policy,omitempty"`
+
+    // ContainerPort specifies the port which will be used by the image container 
+	ContainerPort int `json:"container-port,omitempty"`
+
+}
+```
+
+## Design Details
+
+### Test Plan
+
+To ensure this implementation a new project example should be generated in the [testdata](../testdata/) directory of the project. See the [test/testadata/generate.sh](../test/testadata/generate.sh). Also, we should use this scaffold in the [integration tests](../test/e2e/) to ensure that the data scaffolded with works on the cluster as expected.
+
+### Graduation Criteria
+
+- The new plugin will only be support `project-version=3` 
+- The attribute image with the value informed should be added to the resources model in the PROJECT file to let the tool know that the Resource get done with the common basic code implementation: 
+
+```yaml
+plugins:
+    deploy-image.go.kubebuilder.io/v1beta1:
+        resources:
+          - domain: example.io
+            group: crew
+            kind: Captain
+            version: v1
+            image: "<some-registry>/<project-name>:<tag>
+``` 
+
+For further information check the definition agreement register in the comment https://github.com/kubernetes-sigs/kubebuilder/issues/1941#issuecomment-778649947. 
+
+## Open Questions 
+
+1. Should we allow to scaffold the code for an API that is already created for the project? 
+No, at least in the first moment to keep the simplicity.  
+
+2. Should we support StatefulSet and Deployments?
+The idea is we start it by using a Deployment. However, we can improve the feature in follow-ups to support more default types of scaffolds which could be like `kubebuilder create api --group=crew --version=v1 --image=myexample:0.0.1 --kind=App --plugins=deploy-image.go.kubebuilder.io/v1beta1 --type=[deployment|statefulset|webhook]`
+
+3. Could this feature be useful to other languages or is it just valid to Go based operators?
```

Comment:
Since kubebuilder is only concerned with Go projects and the proposal discusses Go-only changes, I think the answer to this question is simply "no".

Reply:
I would like to leave the answer to clarify some questions raised. This plugin is only valid for go. However, the same idea could be used as an example.